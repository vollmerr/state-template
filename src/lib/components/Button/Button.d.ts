import * as React from 'react';

import * as Icon from '../Icon/Icon';

export interface ButtonProps {
  /** Use style variant */
  variant?:  ''
    | 'default'
    | 'primary'
    | 'secondary'
    | 'highlight'
    | 'standout';

  /** Style as active */
  active?: boolean;

  /** Base button class to apply */
  baseClass?: string;

  /** Icon to render */
  iconProps: Icon.IconProps;

  /** HTML tag to render as */
  tag?: React.ReactNode;

  /** Content to render, mutually exlusive with `children` */
  text?: React.ReactNode;

  /** Button type */
  type?: ''
    | 'button'
    | 'submit'
    | 'reset';

  /** Style class name to attach to button */
  className?: string;

  /** HTML tag to render as */
  tag?: string | React.ReactNode;

  /** Content to render */
  children?: React.ReactNode;

  /** Icon to render */
  iconProps: Icon.IconProps;
}

declare class Button extends React.Component<ButtonProps, {}> {}

export default Button;
