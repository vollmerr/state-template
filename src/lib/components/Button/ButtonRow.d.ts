import * as React from 'react';

export interface ButtonRowProps {
  /** Style class name to attach to button */
  className?: string;

  /** Content to render */
  children?: React.ReactNode;
}

declare class ButtonRow extends React.Component<ButtonRowProps, {}> {}

export default ButtonRow;
