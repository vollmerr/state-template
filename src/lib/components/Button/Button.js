import React from 'react';
import classNames from 'classnames';
import T from 'prop-types';

import Icon from '../Icon';

/**
 * Buttons come in various styles and inherit
 * all properties of a standard html button.
 */
const Button = (props) => {
  const {
    variant,
    active,
    text,
    className,
    tag: Tag,
    children,
    iconProps,
    ...rest
  } = props;

  const btnClass = classNames([
    'btn',
    className,
    { active, [`btn-${variant}`]: variant },
  ]);

  return (
    <Tag data-test={'button'} className={btnClass} {...rest}>
      {iconProps && <><Icon {...iconProps} />{' '}</>}
      {text || children}
    </Tag>
  );
};

Button.propTypes = {
  /** Use style variant */
  variant: T.oneOf([
    '',
    'default',
    'primary',
    'secondary',
    'highlight',
    'standout',
  ]),

  /** Style as active */
  active: T.bool,

  /** Content to render, mutually exclusive with `children` */
  text: T.node,

  /** Button type */
  type: T.oneOf([
    '',
    'button',
    'submit',
    'reset',
  ]),

  /** Style class name to attach to button */
  className: T.string,

  /** HTML tag to render as */
  tag: T.oneOfType([T.func, T.string]),

  /** Content to render */
  children: T.node,

  /** Icon to render */
  iconProps: T.shape({ ...Icon.propTypes }),
};

Button.defaultProps = {
  variant: '',
  active: false,
  text: null,
  type: 'button',
  className: '',
  tag: 'button',
  children: null,
  iconProps: null,
};

export default Button;
