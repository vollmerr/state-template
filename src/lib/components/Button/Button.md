```jsx
import { ButtonRow } from 'state-template';

<ButtonRow>
  <Button text={'button'} />
  <Button text={'default button'} variant={'default'} />
  <Button text={'primary button'} variant={'primary'} />
  <Button text={'secondary button'} variant={'secondary'} />
  <Button text={'highlight button'} variant={'highlight'} />
  <Button text={'standout button'} variant={'standout'} />
</ButtonRow>
```

### active

```jsx
import { ButtonRow } from 'state-template';

<ButtonRow>
  <Button text={'button'} />
  <Button text={'default button'} variant={'default'} active />
  <Button text={'primary button'} variant={'primary'} active />
  <Button text={'secondary button'} variant={'secondary'} active />
  <Button text={'highlight button'} variant={'highlight'} active />
  <Button text={'standout button'} variant={'standout'} active />
</ButtonRow>
```

### disabled

```jsx
import { ButtonRow } from 'state-template';

<ButtonRow>
  <Button text={'button'} disabled />
  <Button text={'default button'} variant={'default'} disabled />
  <Button text={'primary button'} variant={'primary'} disabled />
  <Button text={'secondary button'} variant={'secondary'} disabled />
  <Button text={'highlight button'} variant={'highlight'} disabled />
  <Button text={'standout button'} variant={'standout'} disabled />
</ButtonRow>
```

### with icon

```jsx
import { ButtonRow } from 'state-template';

const iconProps = {
  name: 'plus-mark',
};

<ButtonRow>
  <Button text={'button'} iconProps={iconProps} />
  <Button text={'default button'} variant={'default'} iconProps={iconProps} />
  <Button text={'primary button'} variant={'primary'} iconProps={iconProps} />
  <Button text={'secondary button'} variant={'secondary'} iconProps={iconProps} />
  <Button text={'highlight button'} variant={'highlight'} iconProps={iconProps} />
  <Button text={'standout button'} variant={'standout'} iconProps={iconProps} />
</ButtonRow>
```

### passed props
Other properties get passed down to the button component, such as onChange handlers, id, etc.

```jsx
import { ButtonRow } from 'state-template';

<ButtonRow>
  <Button text={'click me'} onClick={() => alert('You Clicked Me!!')} variant={'default'} />
  <Button text={'other props'} id={'ButtonId'} style={{ fontSize: '20px' }} variant={'primary'} />
</ButtonRow>
```
