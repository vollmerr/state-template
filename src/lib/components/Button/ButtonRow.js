import React from 'react';
import T from 'prop-types';
import classNames from 'classnames';

// Renders buttons in a row with spacing between
const ButtonRow = ({ className, children, ...rest }) => {
  const cn = classNames([
    'btn-row',
    className,
  ]);

  return (
    <div data-test={'button-row'} className={cn} {...rest}>
      {children}
    </div>
  );
};

ButtonRow.propTypes = {
  className: T.string,
  children: T.node.isRequired,
};

ButtonRow.defaultProps = {
  className: '',
};

export default ButtonRow;
