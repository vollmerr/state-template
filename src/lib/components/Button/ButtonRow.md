```jsx
<ButtonRow>
  <Button text={'button 1'} />
  <Button text={'button 2'} />
  <Button text={'button 3'} />
</ButtonRow>
```
