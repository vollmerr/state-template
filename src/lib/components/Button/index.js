import React from 'react';

const Button = (props) => {
  const {
    primary,
    highlight,
    secondary,
    standout,
    text,
    type,
    className,
    ...btnProps
  } = props;

  let btnStyle = 'default';
  if (primary) {
    btnStyle = 'primary';
  } else if (highlight) {
    btnStyle = 'highlight';
  } else if (secondary) {
    btnStyle = 'secondary';
  } else if (standout) {
    btnStyle = 'standout';
  }

  const btnType = type || 'button';
  const btnClass = `btn btn-${btnStyle} ${className}`;

  // eslint-disable-next-line react/button-has-type
  return <button type={btnType} className={btnClass} {...btnProps}>{text}</button>;
};

export default Button;
