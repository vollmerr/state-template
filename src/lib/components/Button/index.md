### default

```jsx
<div className={'btn-row'}>
  <Button text={'default button'} />
  <Button text={'primary button'} primary />
  <Button text={'secondary button'} secondary />
  <Button text={'highlight button'} highlight />
  <Button text={'standout button'} standout />
</div>
```

### active

```jsx
<div className={'btn-row'}>
  <Button text={'default button'} isActive />
  <Button text={'primary button'} primary isActive />
  <Button text={'secondary button'} secondary isActive />
  <Button text={'highlight button'} highlight isActive />
  <Button text={'standout button'} standout isActive />
</div>
```

### disabled

```jsx
<div className={'btn-row'}>
  <Button text={'default button'} disabled />
  <Button text={'primary button'} primary disabled />
  <Button text={'secondary button'} secondary disabled />
  <Button text={'highlight button'} highlight disabled />
  <Button text={'standout button'} standout disabled />
</div>
```

### passed props
Other properties get passed down to the button component, such as onChange handlers, id, etc.

```jsx
<div className={'btn-row'}>
  <Button text={'click me'} onClick={() => alert('You Clicked Me!!')} />
  <Button text={'other props'} id={'ButtonId'} style={{ fontSize: '20px' }} />
</div>
```
