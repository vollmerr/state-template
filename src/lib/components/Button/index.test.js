import React from 'react';
import { shallow } from 'enzyme';

import { Button as fromLib } from '../..';
import Icon from '../Icon';
import Button from '.';

const props = {};

let wrapper;
describe('Button', () => {
  beforeEach(() => {
    wrapper = shallow(<Button {...props}>children...</Button>);
  });

  it('should be exported as part of the library', () => {
    expect(fromLib).toEqual(Button);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render shorthand `text`', () => {
    wrapper.setProps({ text: 'test text' });
    expect(wrapper.text()).toBe('test text');
  });

  it('should render an icon if passed `iconProps`', () => {
    const iconProps = { name: 'test' };
    wrapper.setProps({ iconProps });
    expect(wrapper.find(Icon).length).toBe(1);
  });

  it('should allow `variant`s', () => {
    wrapper.setProps({ variant: 'secondary' });
    const className = wrapper.find('button').prop('className');
    expect(className).toMatch(/btn-secondary/);
  });

  it('should attach `className`s', () => {
    wrapper.setProps({ className: 'textClass' });
    const className = wrapper.find('button').prop('className');
    expect(className).toMatch(/textClass/);
  });

  it('should allow different `tag`s', () => {
    wrapper.setProps({ tag: 'a' });
    expect(wrapper.find('a').length).toBe(1);
  });

  it('should pass additional props', () => {
    wrapper.setProps({ id: 'testId', type: 'submit' });
    const button = wrapper.find('button');
    expect(button.prop('id')).toBe('testId');
    expect(button.prop('type')).toBe('submit');
  });
});
