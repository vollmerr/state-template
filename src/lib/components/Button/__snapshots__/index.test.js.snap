// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render correctly 1`] = `
<button
  className="btn"
  type="button"
>
  test text
</button>
`;
